src/k8s-engine/client.js

```diff
@@ -48,173 +47,321 @@ class ErrorSuppressingStdin extends events.EventEmitter {
+    return super.has(`${namespace}/${endpoint}:${port}`);
+  }
+
+  /**
+   * Iterate through the entries.
+   * @returns {IterableIterator<[string, string, number, net.Server]>}
+   */
+  *[Symbol.iterator]() {
+    const iter = super[Symbol.iterator]();
+    for (const [key, server] of iter) {
+      const [namespace, endpoint, port] = /^([^/]+)\/([^:]+):(\d+)$/.exec(key);
+      yield [namespace, endpoint, parseInt(port), server];
+    }
+  }
+}
+
 /**
  * KubeClient is a Kubernetes client that will _only_ manage the cluster we spin
  * up internally.  The user should call initialize() once the cluster has been
  * created.
  */
-class KubeClient {
-    #kubeconfig = new k8s.KubeConfig();
-    /**
-     * @type k8s.PortForward?
-     */
-    #forwarder = null;
-    /**
-     * @type net.Server?
-     */
-    #server = null;
-    #shutdown = false;
+class KubeClient extends events.EventEmitter {
+  #kubeconfig = new k8s.KubeConfig();
+  /**
+   * @type k8s.PortForward?
+   */
+  #forwarder = null;
 
-    /**
-     * initialize the KubeClient so that we are ready to talk to it.
-     */
-    constructor() {
-      this.#kubeconfig.loadFromDefault();
-      this.#kubeconfig.currentContext = 'rancher-desktop';
-      this.#forwarder = new k8s.PortForward(this.#kubeconfig, true);
-      this.#shutdown = false;
-    }
+  #shutdown = false;
 
-    // Notify that the client the underlying Kubernetes cluster is about to go
-    // away, and we should remove any pending work.
-    destroy() {
-      this.#shutdown = true;
-      this.#server?.close();
-      this.#server = null;
-    }
+  /**
+   * Kubernetes services across all namespaces.
+   * @type {k8s.ListWatch<k8s.V1Service>}
+   */
+  #services;
 
-    /**
-     * @type k8s.CoreV1Api
-     */
-    get #coreV1API() {
-      this.#_coreV1API ||= this.#kubeconfig.makeApiClient(k8s.CoreV1Api);
-      return this.#_coreV1API;
+  /**
+   * Active port forwarding servers.  This records the desired state: if an
+   * entry exists, then we want to set up port forwarding for it.
+   */
+  #servers = new ForwardingMap();
+
+  /**
+   * initialize the KubeClient so that we are ready to talk to it.
+   */
+  constructor() {
+    super();
+    this.#kubeconfig.loadFromDefault();
+    this.#kubeconfig.currentContext = 'rancher-desktop';
+    this.#forwarder = new k8s.PortForward(this.#kubeconfig, true);
+    this.#shutdown = false;
+
+    // Set up a watch for services
+    // Since the watch API we have _doesn't_ notify us when things have
+    // changed, we'll need to do some trickery and wrap the underlying watcher
+    // with our own code.
+    const watch = new k8s.Watch(this.#kubeconfig);
+    const wrappedCallback = (callback, ...args) => {
+      callback(...args);
+      this.emit('service-changed', this.listServices());
+    };
+    const wrappedWatch = {
+      watch(path, queryParams, callback, ...extras) {
+        watch.watch(path, queryParams, wrappedCallback.bind(this, callback), ...extras);
```

Comment:
FWICT, you're binding `wrappedCallback` to the `wrappedWatch` object. Or does the callback resolve `this.emit` correctly?

Reply:
`wrappedCallback` is an arrow function, so it's already bound; the `.bind()` call here is only just for currying (that is, adding the `callback` argument to the function).